## Contents
    01.Importing libraries
    02.Importing Data
    03.Data Wrangling
    04.Data Consistency Checks
    05.Merging Dataframes

# 01. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02.Importing Data

In [2]:
#setting path
path=r'C:\Users\rache\Documents\Instacart Basket Analysis'

In [3]:
#importing customers.csv
cust=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [14]:
cust.shape

(206209, 10)

# 03.Data Wrangling

### Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [7]:
#renaming surnam column to be consistent with first name column
cust.rename(columns = {'Surnam':'Last Name'}, inplace = True)

In [8]:
#renaming STATE column to State
cust.rename(columns = {'STATE':'State'}, inplace = True)

In [12]:
#renaming n_dependants column to be more explanatory and correcting spelling error
cust.rename(columns = {'n_dependants':'No. of Dependents'}, inplace = True)

In [10]:
#renaming fam_status column to be more clear
cust.rename(columns = {'fam_status':'family_status'}, inplace = True)

In [13]:
cust.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   No. of Dependents family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

# 04.Data Consistency Checks

### Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [20]:
cust.describe()

Age  No. of Dependents         income
count  206209.000000      206209.000000  206209.000000
mean       49.501646           1.499823   94632.852548
std        18.480962           1.118433   42473.786988
min        18.000000           0.000000   25903.000000
25%        33.000000           0.000000   59874.000000
50%        49.000000           1.000000   93547.000000
75%        66.000000           3.000000  124244.000000
max        81.000000           3.000000  593901.000000

In [18]:
#Checking data types
cust.dtypes

user_id              object
First Name           object
Last Name            object
Gender               object
State                object
Age                   int64
date_joined          object
No. of Dependents     int64
family_status        object
income                int64
dtype: object

In [19]:
cust['user_id'] = cust['user_id'].astype('str')

In [21]:
#Checking for missing values
cust.isnull().sum()

user_id                  0
First Name           11259
Last Name                0
Gender                   0
State                    0
Age                      0
date_joined              0
No. of Dependents        0
family_status            0
income                   0
dtype: int64

In [22]:
cust_nan=cust[cust['First Name'].isnull()==True]

In [23]:
cust_nan

user_id First Name Last Name  Gender         State  Age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        No. of Dependents family_status  income  
53                      2       married   41709  
73                      0        single   82518  
82                      3       married  117099  
99                      1       married  155673  
105                     3       married  151819  
...                   ...           ...     ...  
206038                  3       married   87783  
206044                  2       married  108488  
206090                  1       married   45275  
206162                  0        single   56325  
206171                  2       married   59222  

[11259 rows x 10 columns]

#### Having 11,259 missing first names should not have much impact as we have all user ids to link dataframes.  Missing values can remain.

In [24]:
# Checking for mixed data types 
for col in cust.columns.tolist():
  weird = (cust[[col]].applymap(type) != cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (cust[weird]) > 0:
    print (col)

First Name


#### First Name column has mixed data type

In [25]:
cust['First Name'] = cust['First Name'].astype('str')

In [27]:
# Checking for duplicates
cust[cust.duplicated()]

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, No. of Dependents, family_status, income]
Index: []

#### No duplicates

In [28]:
#reducing data types to save RAM before merge
cust['Age'] = cust['Age'].astype('int8')
cust['No. of Dependents'] = cust['No. of Dependents'].astype('int8')
cust['income'] = cust['income'].astype('int32')

# 05.Merging Dataframes

### Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type! 

In [29]:
#Importing orders_products_merged_ver2.pkl
ords_prods=pd.read_pickle(os.path.join(path, '02 Data','Prepared Data','orders_products_merged_ver2.pkl'))

In [30]:
#Checking size of current dataframe
ords_prods.shape

(32404859, 20)

In [31]:
#Checking data type for key column user_id
ords_prods.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_last_order     float64
first_time_customer          bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
_merge                   category
Unnamed: 0                  int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
dtype: object

In [32]:
#Changing user_id column to string/object
ords_prods['user_id'] = ords_prods['user_id'].astype('str')

In [33]:
#verifying change
ords_prods.dtypes

order_id                    int64
user_id                    object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_last_order     float64
first_time_customer          bool
product_id                  int64
add_to_cart_order           int64
reordered                   int64
_merge                   category
Unnamed: 0                  int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
dtype: object

In [36]:
#merging cust df into ords_prods df using user_id key column
ords_prods_cust=ords_prods.merge(cust, on ='user_id')

In [37]:
ords_prods_cust

order_id user_id  order_number  order_day_of_week  \
0          2539329       1             1                  2   
1          2398795       1             2                  3   
2           473747       1             3                  3   
3          2254736       1             4                  4   
4           431534       1             5                  4   
...            ...     ...           ...                ...   
32404854    156685  106143            26                  4   
32404855    484769   66343             1                  6   
32404856   1561557   66343             2                  1   
32404857    276317   66343             3                  6   
32404858   2922475   66343             4                  1   

          order_hour_of_day  days_since_last_order  first_time_customer  \
0                         8                    NaN                 True   
1                         7                   15.0                False   
2                        12                   21.0                False   
3                         7                   29.0                False   
4                        15                   28.0                False   
...                     ...                    ...                  ...   
32404854                 23                    5.0                False   
32404855                 11                    NaN                 True   
32404856                 11                   30.0                False   
32404857                 15                   19.0                False   
32404858                 12                   30.0                False   

          product_id  add_to_cart_order  reordered  ... busiest_period_of_day  \
0                196                  1          0  ...        Average orders   
1                196                  1          1  ...        Average orders   
2                196                  1          1  ...           Most orders   
3                196                  1          1  ...        Average orders   
4                196                  1          1  ...           Most orders   
...              ...                ...        ...  ...                   ...   
32404854       19675                  1          1  ...        Average orders   
32404855       47210                  1          0  ...           Most orders   
32404856       47210                  1          1  ...           Most orders   
32404857       47210                  1          1  ...           Most orders   
32404858       47210                  1          1  ...           Most orders   

          First Name Last Name  Gender      State  Age date_joined  \
0              Linda    Nguyen  Female    Alabama   31   2/17/2019   
1              Linda    Nguyen  Female    Alabama   31   2/17/2019   
2              Linda    Nguyen  Female    Alabama   31   2/17/2019   
3              Linda    Nguyen  Female    Alabama   31   2/17/2019   
4              Linda    Nguyen  Female    Alabama   31   2/17/2019   
...              ...       ...     ...        ...  ...         ...   
32404854      Gerald     Yates    Male     Hawaii   25   5/26/2017   
32404855  Jacqueline    Arroyo  Female  Tennessee   22   9/12/2017   
32404856  Jacqueline    Arroyo  Female  Tennessee   22   9/12/2017   
32404857  Jacqueline    Arroyo  Female  Tennessee   22   9/12/2017   
32404858  Jacqueline    Arroyo  Female  Tennessee   22   9/12/2017   

         No. of Dependents family_status income  
0                        3       married  40423  
1                        3       married  40423  
2                        3       married  40423  
3                        3       married  40423  
4                        3       married  40423  
...                    ...           ...    ...  
32404854                 0        single  53755  
32404855                 3       married  46151  
32404856                 3       married  46151  
32404857                 3       marri

#### Checking 9 new columns were added. Total of 10 in cust df but user_id column was key and not added again

### Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [38]:
#Exporting new dataframe
ords_prods_cust.to_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'full_instacart_data.pkl'))